Don't flatten hierarchy when installing headers

tkoolen opened the pull request:
Ran into some very annoying issues while updating DRC's version of Drake, which I'd finally like to resolve for good.

Until now, (almost) all drake headers were installed to a single directory (build/include/drake), even though they exist in a directory hierarchy in the source tree. This resulted in all kinds of trouble when an external project like DRC referred to the headers in their installed location, as references between the headers that worked in the source tree were no longer valid after flattening. The only thing that allowed this to work was that we used include_directories all over the place, so that headers could be included without a reference to their path in the source tree. An additional issue was that headers could overwrite other installed headers with the same name existing in a different directory of the source tree (e.g. DrakeCollision::Element and DrakeShapes::Element), necessitating an ad-hoc solution.

This pull request fixes these issues by:
* replicating the directory structure in the source tree in build/include/drake
* removing all of the include_directories calls from all but the top level CMakeLists.txt (cleans things up quite a bit)
* only using paths relative to drake or relative to the current file when including headers.

In addition, the export headers are now also installed into build/include/drake.

rdeits commented:
Hooray!!

RussTedrake commented:
oh @tkoolen .  why must all of your pull requests touch > 200 files?  :)

RussTedrake commented on drake/CMakeLists.txt:
```diff
@@ -42,13 +47,15 @@ function(add_library_with_exports)
+  install(FILES ${exports_abs_path} DESTINATION include/drake/)
 endfunction()
-include_directories(${PROJECT_BINARY_DIR}/exports)
+include_directories(${PROJECT_BINARY_DIR}/exports/)
```
just to be clear... internally, compilation using "drake/libname_export.h" calls "pod-build/exports/drake/libname_export.h", but someone referencing build/include will get it from build/include/drake/libname_export.h .  it's a little complicated, but I guess that works.  perhaps add a line of textual documentation around this method?

RussTedrake replied:
add BEFORE here, too

RussTedrake commented on drake/CMakeLists.txt:
```diff
@@ -125,7 +134,9 @@ file(WRITE ${PROJECT_SOURCE_DIR}/core/Path.h
-
+set_directory_properties(PROPERTIES ADDITIONAL_MAKE_CLEAN_FILES ${drake_path_header})
+install(FILES ${drake_path_header} DESTINATION include/drake/)
+include_directories(${generated_dir})
```
same trick is happening here, but with `pod-build/generated/drake/Path.h` which gets installed to `build/include/drake/Path.h` for easy referencing later...

RussTedrake replied:
BEFORE

RussTedrake commented on drake/CMakeLists.txt:
```diff
@@ -42,13 +47,15 @@ function(add_library_with_exports)
 endfunction()
-include_directories(${PROJECT_BINARY_DIR}/exports)
+include_directories(${PROJECT_BINARY_DIR}/exports/)
+include_directories(${PROJECT_SOURCE_DIR}/..)
```
i think we have to be very careful here.  if I include, e.g., `drake/test/testUtil.h` (the first example below), then it's actually going to get the previously installed version of that file, not the file in the source directory.  Because `build/include` is already put into the include_directories from pods.cmake (if I remember correctly).  

i think you want this line to be 
```
include_directories(BEFORE ${PROJECT_SOURCE_DIR}/..)
```
(see https://cmake.org/cmake/help/v3.0/command/include_directories.html )

even then, it's a bit of a slippery slope.  but i guess that's probably the right thing.

RussTedrake commented on drake/CMakeLists.txt:
```diff
@@ -203,7 +214,7 @@ else()
 endif()
 
-include_directories(thirdParty/spruce/include)
+include_directories(thirdParty/spruce/include) # FIXME: should be in build/include
```
i don't agree with the fixme.  i think we should only reference prerequisite externals in build/include, not anything contained inside of drake.

RussTedrake commented:
ok.  i agree this is much better.  but if i'm write, there is a major bug in the order of includes (described above).

patmarion commented:
I like this change.  A few things:

  - as Russ pointed out, since the install destination is added as include dirs, be careful about include orders
  - i think including ${PROJECT_SOURCE_DIR}/.. looks sketchy.  Perhaps it's a larger change than you might want, but many projects will embed the full layout in source.  Something like this for each module, there are many alternative organizations, but here's an example:

drake/core/include/drake/core/public_header.h
drake/core/src/file.cpp
drake/core/src/private_header.h


I'm not excited about very long hierarchy strings, for example:

"drake/systems/plants/joints/DrakeJoint.h"


After this pull request, it might be a time to discuss source code layout.  By installing the headers into a hierarchy, downstream projects must use the hierarchy so it's nice if it's got a consistent logic in the layout design.

One way to think of the directory hierarchy is defining a module layout.  A lot of c++ project are organized into modules.  Each module has a name, a library file, and public headers.  Currently drake has a folder layout that sort of looks like modules (and submodules) but the libraries created in each directory don't really reflect the directory name.  For example, the solvers directory doesn't actually create a library named drakeSolvers.  The util directory does create drakeUtil, but it also creates other libraries, like drakePolynomial.  If you do a module design, then you can have a core module, and the next module depends on core, and you can start documenting module dependencies and things like that.

Where you have systems/plants/joints, and classes in systems/plants depend on classes in joints, then maybe joints can be a new module dependency of systems/plants, rather than having joints be a subdirectory.

tkoolen commented:
Addressed the comments. Thanks for the review.

I think 239 files is a record for me, but then again this pull request only added 476 lines and deleted 605, it only tries to accomplish a single goal, and it changes things for the better.

> After this pull request, it might be a time to discuss source code layout.

Yep, sounds good.

tkoolen commented:
build errors are not my fault, I believe.

rdeits commented:
Great! This also cleaned up some of the Swig binding files a bit, which previously had lots of `include_directories` calls (see #1629 )
